amaranth/hdl/_ir.py

```diff
@@ -551,7 +551,7 @@ def _assign_port_names(self):
         new_ports = []
         assigned_names = {name for name, conn, dir in self.ports if name is not None}
         for name, conn, dir in self.ports:
-            if name is None:
+            if name is None and conn.name != "": # Private signals have no name.
```

Comment:
if `conn.name == ""` we need to fail here entirely — we cannot have a top-level port with no name

Reply:
The port’s name here is None, we are stopping it from getting the non-name of the signal. If None it should be given a name later.

Reply:
but there is no "later" here — when this function runs, the top-level ports are finalised, and they must have a name

this is different from non-top-level ports, where we have auto-naming. top-level ports are the *interface* of the design, and we cannot allow unnamed stuff there